### PR TITLE
Adds support for gogs-client to retrieve repos by username and org

### PR DESCRIPTION
**Pull request by kmadel**
This is related to [gogs/gogs-client issue 37](https://github.com/gogits/go-gogs-client/issues/37)

Requires [gogs/gogs-client PR 45](https://github.com/gogits/go-gogs-client/pull/45)

Provides the ability to retrieve repos by username or by org via the REST API (gogs-client).

### CODE REVIEW

**Comment by kmadel** [comment]
By the way, just wanted to include a link to the GitHub documentation:

- https://developer.github.com/v3/repos/#list-user-repositories
- https://developer.github.com/v3/repos/#list-organization-repositories

File: routers/api/v1/api.go [hunk 1]
```diff
@@ -215,6 +215,9 @@ func RegisterRoutes(m *macaron.Macaron) {
 		}, reqToken())
 
 		// Repositories
+		m.Get("/users/:username/repos", repo.ListUserRepos)
+		m.Get("/orgs/:org/repos", repo.ListOrgRepos)
+
 		m.Combo("/user/repos", reqToken()).Get(repo.ListMyRepos).
```
Comment (Unknwon) [inline routers/api/v1/api.go]: Why not put two other routes after this line (221)?

File: routers/api/v1/repo/repo.go [hunk 2]
```diff
@@ -109,6 +108,51 @@ func ListMyRepos(ctx *context.APIContext) {
...
 
+// https://github.com/gogits/go-gogs-client/wiki/Repositories#list-your-repositories
+func ListMyRepos(ctx *context.APIContext) {
+	listUserRepos(ctx, ctx.User, true)
+}
+
+// https://github.com/gogits/go-gogs-client/wiki/Repositories#list-user-repositories
+func ListUserRepos(ctx *context.APIContext) {
+	u, err := models.GetUserByName(ctx.Params(":username"))
+
```
Comment (Unknwon) [inline routers/api/v1/repo/repo.go]: Remove this empty line.

File: routers/api/v1/repo/repo.go [hunk 3]
```diff
@@ -109,6 +108,51 @@ func ListMyRepos(ctx *context.APIContext) {
...
+	if ctx.Written() {
+		return
+	}
+	listUserRepos(ctx, u, false)
+}
+
+// https://github.com/gogits/go-gogs-client/wiki/Repositories#list-org-repositories
+func ListOrgRepos(ctx *context.APIContext) {
+	u, err := models.GetUserByName(ctx.Params(":org"))
+
```
Comment (Unknwon) [inline routers/api/v1/repo/repo.go]: Remove this empty line.

File: routers/api/v1/repo/repo.go [hunk 4]
```diff
@@ -109,6 +108,51 @@ func ListMyRepos(ctx *context.APIContext) {
...
+// https://github.com/gogits/go-gogs-client/wiki/Repositories#list-user-repositories
+func ListUserRepos(ctx *context.APIContext) {
+	u, err := models.GetUserByName(ctx.Params(":username"))
+
+	if err != nil {
+		if models.IsErrUserNotExist(err) {
+			ctx.Status(404)
+		} else {
+			ctx.Error(500, "ListUserRepos", err)
+		}
```
Comment (Unknwon) [inline routers/api/v1/repo/repo.go]: `return`

File: routers/api/v1/repo/repo.go [hunk 5]
```diff
@@ -109,6 +108,51 @@ func ListMyRepos(ctx *context.APIContext) {
...
+	if !u.IsOrganization() {
+		ctx.Status(404)
+	}
+
+	if err != nil {
+		if models.IsErrUserNotExist(err) {
+			ctx.Status(404)
+		} else {
+			ctx.Error(500, "ListOrgRepos", err)
+		}
```
Comment (Unknwon) [inline routers/api/v1/repo/repo.go]: `return`

File: routers/api/v1/repo/repo.go [hunk 6]
```diff
@@ -109,6 +108,51 @@ func ListMyRepos(ctx *context.APIContext) {
...
+	}
+
+	if ctx.Written() {
```
Comment (Unknwon) [inline routers/api/v1/repo/repo.go]: Redundant.

File: routers/api/v1/repo/repo.go [hunk 7]
```diff
@@ -109,6 +108,51 @@ func ListMyRepos(ctx *context.APIContext) {
...
+
+	if err != nil {
+		if models.IsErrUserNotExist(err) {
+			ctx.Status(404)
+		} else {
+			ctx.Error(500, "ListUserRepos", err)
+		}
+	}
+
+	if ctx.Written() {
```
Comment (Unknwon) [inline routers/api/v1/repo/repo.go]: Redundant.

**Comment by kmadel** [comment]
I believe all the line notes have been addressed - wasn't certain about "Why not put two other routes after this line (221)?" - but hopefully that is what you meant. Note the other top level repositories entries refer to `user` and `org` and the two new GETs refer to `users` and `orgs` - I'm happy to move where ever you believe it makes most sense.

Thanks for the feedback!

File: routers/api/v1/api.go [hunk 8]
```diff
@@ -215,12 +215,11 @@ func RegisterRoutes(m *macaron.Macaron) {
 		}, reqToken())
 
 		// Repositories
-		m.Get("/users/:username/repos", repo.ListUserRepos)
-		m.Get("/orgs/:org/repos", repo.ListOrgRepos)
-
 		m.Combo("/user/repos", reqToken()).Get(repo.ListMyRepos).
 			Post(bind(api.CreateRepoOption{}), repo.Create)
 		m.Post("/org/:org/repos", reqToken(), bind(api.CreateRepoOption{}), repo.CreateOrgRepo)
```
Comment (Unknwon) [inline routers/api/v1/api.go]: Looks like we can use `m.Combo` here?

Comment (kmadel) [inline routers/api/v1/api.go]: Hmm, not sure because one is `/users/...` and the other is `/orgs/...`

Comment (Unknwon) [inline routers/api/v1/api.go]: Ohh... you're right, sorry.

File: routers/api/v1/repo/repo.go [hunk 9]
```diff
@@ -76,9 +76,8 @@ func Search(ctx *context.APIContext) {
 	})
 }
 
-// https://github.com/gogits/go-gogs-client/wiki/Repositories#list-your-repositories
-func ListMyRepos(ctx *context.APIContext) {
-	ownRepos, err := models.GetUserRepositories(ctx.User.ID, true, 1, ctx.User.NumRepos)
+func listUserRepos(ctx *context.APIContext, u *models.User, all bool) {
```
Comment (Unknwon) [inline routers/api/v1/repo/repo.go]: `all` is not used?

File: routers/api/v1/repo/repo.go [hunk 10]
```diff
@@ -76,9 +76,8 @@ func Search(ctx *context.APIContext) {
 	})
 }
 
-// https://github.com/gogits/go-gogs-client/wiki/Repositories#list-your-repositories
-func ListMyRepos(ctx *context.APIContext) {
-	ownRepos, err := models.GetUserRepositories(ctx.User.ID, true, 1, ctx.User.NumRepos)
+func listUserRepos(ctx *context.APIContext, u *models.User) {
+	ownRepos, err := models.GetUserRepositories(u.ID, true, 1, u.NumRepos)
```
Comment (Unknwon) [inline routers/api/v1/repo/repo.go]: This does not look right, you're listing all private repositories..

Comment (kmadel) [inline routers/api/v1/repo/repo.go]: You are correct - I will update the PR - hopefully tonight. Thanks!

**Pull request closed without merging**